Add support for .env file

kinghuang opened the pull request:
Add support for using a .env file placed in the folder docker-compose is exected from (the current working directory) to set environment variables. Also enable the use of the COMPOSE_ prefixed CLI environment variables to set the Compose project name and Compose file(s).

Resolves #5.

Note that this is based on issue-7-multiple-compose-files.

kinghuang committed:
Set action='append' on --file argument

Set action=‘append’ when adding the argument parser for --file (-f), so
that it can accumulate multiple files.

If the length of args.file is 0 (i.e., no file was specified), then
attempt to open docker-compose.yml as the file. docker-compose.yml is
no longer set as the argument's default value  because action=append
adds to it instead of overriding it.

For now, the first (0-index) file is used to determine compose_base_dir
and create the DockerCompose object.

kinghuang committed:
Merge multiple compose files together

Merge all the compose files in args.file together into one compose file
before passing it on to the DockerCompose constructor. This allows
multiple compose files to be used together to form a single
configuration.

Merge function based on
http://stackoverflow.com/questions/7204805/dictionaries-of-dictionaries-
merge/7205107#7205107, and modified to also extend lists.

Resolves #7.

kinghuang committed:
Read in .env file from current working directory

Docker Compose supports declaring default environment variables in a
file named .env in the current working directory. Attempt to find such
a file and read it in.

kinghuang committed:
Declare default CLI environment variables

Initialize env with the default Docker Compose CLI environment
variables (starting with COMPOSE_). If an .env file is specified, it
will override any values in the default env.

kinghuang committed:
Use COMPOSE_FILE if --file is not specified

If no file is specified, use the CLI environment variable COMPOSE_FILE
to determine which compose files to load. This defaults to
docker-compose.yml.

kinghuang committed:
Load .env into os.environ instead of env dict

Set default environment variables and load declarations in .env into
the environment (os.environ) instead of a local dict. That way the full
environment is considered, not just defaults specified in this script
or loaded from .env.

This also supports the use of subprocess.Popen later on for variable
substitution using environment variables.

kinghuang committed:
Set --project-name default to COMPOSE_PROJECT_NAME

Use the value of the COMPOSE_PROJECT_NAME environment variable as the
default value for the --project-name argument. This allows the project
name to be set from the environment or .env file, matching Docker
Compose.

ddrozdov commented on docker_compose_swarm_mode.py:
```diff
@@ -214,9 +214,23 @@ def start(self, services=None):
+
+    if os.path.isfile(env_path):
+        with open(env_path) as env_file:
+            map(lambda e: os.environ.update({e[0]:e[1]}), (line.strip().split('=', 1) for line in env_file if not line.startswith('#') and line.strip()))
```
According to the [docs](https://docs.docker.com/compose/env-file/):

> Values present in the environment at runtime will always override those defined inside the .env file

So we should only call `update` if there was no such a key.

I guess your idea was to make `.env` override the initial `COMPOSE_FILE`, ... map.
Then we should first merge that map with the content of the `.env` file and after that update the environment.

kinghuang replied:
Good call, I missed that in the docs! I'll limit the updates from .env to only if the key is not already in os.environ, and move the remaining defaults down.

kinghuang committed:
Only use .env variables when undeclared in os.environ

From the docs: “Values present in the environment at runtime will
always override those defined inside the .env file.”

Adjust the map so that os.environ is only updated if the environment
variable name does not already exist in os.environ. Also, break this up
into two lines so that it doesn't get overly long.

kinghuang committed:
Merge default envs with .env before updating os.environ

Instead of updating os.environ with some default environment variables
then overriding them with .env, merge the two together, then update
os.environ for keys that are not already defined.

kinghuang committed:
Fix iteration over envs dict

Iterate over envs.items() to get both keys and values.

kinghuang commented:
I believe the latest commits should do it. The default envs are merged with .env. Then, os.environ is updated for anything that is not already declared in the environment.
